File: Algorithms/735. Asteroid Collision.ts
function asteroidCollision(asteroids: number[]): number[] {
  const result: number[] = [];
  for (let asteroid of asteroids) {
    if (asteroid > 0 || result.length === 0 || result[result.length - 1] < 0) {
      result.push(asteroid);
    } else {
      let i: number = result.length - 1;
      while (i >= 0) {
        if (result[i] < 0) {
          result.push(asteroid);
          break;
        } else {
          const last: number = result.pop();
          if (last > Math.abs(asteroid)) {
            result.push(last);
            break;
          } else if (last === Math.abs(asteroid)) {
            break;
          } else {
            if (i === 0) {
              result.push(asteroid);
            }
            i--;
          }
        }
      }
    }
  }

  return result;
}

File: Algorithms/735. Asteroid Collision.rs
impl Solution {
    pub fn asteroid_collision(asteroids: Vec<i32>) -> Vec<i32> {
        let mut result: Vec<i32> = Vec::new();

        for &asteroid in asteroids.iter() {
            if asteroid > 0 || result.len() == 0 || result[result.len() - 1] < 0 {
                result.push(asteroid);
            } else {
                let mut i = result.len() - 1;
                while i >= 0 {
                    if result[i] < 0 {
                        result.push(asteroid);
                        break;
                    } else {
                        let last = result.last().unwrap();
                        if *last > asteroid.abs() {
                            break;
                        } else if *last == asteroid.abs() {
                            result.pop();
                            break;
                        } else {
                            result.pop();
                            if i == 0 {
                                result.push(asteroid);
                                break;
                            }
                            i -= 1;
                        }
                    }
                }
            }
        }
        result
    }
}
